ddtrace/profiling/collector/stack.pyx

```diff
@@ -83,64 +83,77 @@ IF UNAME_SYSNAME == "Linux":
             return int(tp.tv_nsec + tp.tv_sec * 10e8)
         PyErr_SetFromErrno(OSError)
 
-    class _ThreadTime(object):
+    cdef class _ThreadTime(object):
+        cdef dict _last_thread_time
+
         def __init__(self):
-            # This uses a tuple of (thread unique id, pthread_id) as the key to identify the thread: you'd think using
+            # This uses a tuple of (pthread_id, thread_native_id) as the key to identify the thread: you'd think using
             # the pthread_t id would be enough, but the glibc reuses the id.
             self._last_thread_time = {}
 
-        def __call__(self, thread_ids):
-            threads_cpu_time = {}
-            for pthread_id, thread_native_id in thread_ids.items():
-                key = (pthread_id, thread_native_id)
+        # Only used in tests
+        def _get_last_thread_time(self):
+            return dict(self._last_thread_time)
+
+        def __call__(self, pthread_ids):
+            cdef list cpu_times = []
+            for pthread_id in pthread_ids:
                 # TODO: Use QueryThreadCycleTime on Windows?
                 # ⚠ WARNING ⚠
                 # `pthread_getcpuclockid` can make Python segfault if the thread is does not exist anymore.
                 # In order avoid this, this function must be called with the GIL being held the entire time.
                 # This is why this whole file is compiled down to C: we make sure we never release the GIL between
                 # calling sys._current_frames() and pthread_getcpuclockid, making sure no thread disappeared.
                 try:
-                    clock_id = p_pthread_getcpuclockid(pthread_id)
+                    cpu_time = p_clock_gettime_ns(p_pthread_getcpuclockid(pthread_id))
                 except OSError:
-                    cpu_time = self._last_thread_time.get(key, 0)
-                else:
-                    try:
-                        cpu_time = p_clock_gettime_ns(clock_id)
-                    except OSError:
-                        cpu_time = self._last_thread_time.get(key, 0)
+                    # Just in case it fails, re-use the old one
+                    # (Note that glibc never fails, it segfaults instead)
+                    cpu_time = 0
```

Comment:
The comment and this line don't match-up, but I'm not sure which we'd prefer 😛

Reply:
Fixed.